File: jdroid-android-core/src/main/java/com/jdroid/android/concurrent/AppExecutors.java
package com.jdroid.android.concurrent;

import android.os.Handler;
import android.os.Looper;

import com.jdroid.java.concurrent.NormalPriorityThreadFactory;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

/**
 * Global executor pools for the whole application.
 *
 * Grouping tasks like this avoids the effects of task starvation
 * (e.g. disk reads don't wait behind webservice requests).
 */
public class AppExecutors {

	private static Executor diskIO = Executors.newSingleThreadExecutor(new NormalPriorityThreadFactory("jdroid-disk"));
	private static Executor networkIO = Executors.newFixedThreadPool(3, new NormalPriorityThreadFactory("jdroid-network"));
	private static Executor mainThread = new MainThreadExecutor();

	public static Executor getDiskIOExecutor() {
		return diskIO;
	}

	public static Executor getNetworkIOExecutor() {
		return networkIO;
	}

	public static Executor getMainThreadExecutor() {
		return mainThread;
	}

	private static class MainThreadExecutor implements Executor {

		private Handler mainThreadHandler = new Handler(Looper.getMainLooper());

		@Override
		public void execute(Runnable command) {
			mainThreadHandler.post(command);
		}
	}
}

File: jdroid-android-core/src/main/java/com/jdroid/android/concurrent/AppExecutors.kt
package com.jdroid.android.concurrent

import android.os.Handler
import android.os.Looper

import com.jdroid.java.concurrent.NormalPriorityThreadFactory

import java.util.concurrent.Executor
import java.util.concurrent.Executors

/**
 * Global executor pools for the whole application.
 *
 * Grouping tasks like this avoids the effects of task starvation
 * (e.g. disk reads don't wait behind webservice requests).
 */
object AppExecutors {

    val diskIOExecutor: Executor = Executors.newSingleThreadExecutor(NormalPriorityThreadFactory("jdroid-disk"))
    val networkIOExecutor: Executor = Executors.newFixedThreadPool(3, NormalPriorityThreadFactory("jdroid-network"))
    val mainThreadExecutor: Executor = MainThreadExecutor()

    private class MainThreadExecutor : Executor {

        private val mainThreadHandler = Handler(Looper.getMainLooper())

        override fun execute(command: Runnable) {
            mainThreadHandler.post(command)
        }
    }
}
